plugin/src/main/java/org/jenkinsci/plugins/structs/describable/DescribableModel.java

```diff
@@ -104,19 +105,39 @@
      */
     private final String[] constructorParamNames;
 
-    /** binds type parameter */
-    static <T> DescribableModel<T> of(Class<T> clazz) {
-        return new DescribableModel<T>(clazz);
+    /** Binds type parameter, preferred means of obtaining a DescribableModel. */
+    public static <T> DescribableModel<T> of(Class<T> clazz) {
+        DescribableModel mod = modelCache.get(clazz.getName());
+        if (mod != null) {
+            return mod;
+        }
+        mod = new DescribableModel<T>(clazz);
+        modelCache.put(clazz.getName(), mod);
+        return mod;
     }
 
+    /** Map class name to cached model. */
+    static ConcurrentHashMap<String, DescribableModel> modelCache = new ConcurrentHashMap<String, DescribableModel>();
+
     /**
      * Loads a definition of the structure of a class: what kind of data
      * you might get back from {@link #uninstantiate} on an instance,
      * or might want to pass to {@link #instantiate(Map)}.
+     *
+     * Use {@link #of(Class)} instead -- that will returned cached instances.
      */
     public DescribableModel(Class<T> clazz) {
         this.type = clazz;
 
+        DescribableModel mod = modelCache.get(clazz.getName());
+        if (mod != null) {
```

Comment:
No, this is wrong, you cannot key by class name alone!

At a minimum you must double check that the recorded model type is `== clazz` before proceeding with the cache.

Reply:
If it's not safe to key by classname then *something* has gone horribly wrong with ClassLoading.

Reply:
No, there is nothing wrong with class loading. You can have two otherwise unrelated `Class`es with the same `name` in different `ClassLoader`s. You must treat them as distinct for all purposes.

Reply:
Right but my understanding was that we *shouldn't* have that (at least not with the same defining classloader) for Jenkins classes -- that's not true for groovyCPS or pipeline-internal classes, but they don't get a DescribableModel anyway. 

Is that incorrect, and can you point me to an example of where this will generate a concrete issue?

Reply:
There is nothing stopping two Jenkins plugins from defining a class by the same name. Not a situation we necessarily _want_, for various practical reasons, but the Jenkins plugin manager supports it.

Reply:
Okay, that makes some sense - I will take you world that the plugin classloader system permits it, although I strongly suspect in practice it will fall down somewhere (probably in many somewheres) if you actually do it.